### PR TITLE
Use torchtext for MP-CNN

### PR DESCRIPTION
**Pull request by tuzhucheng**
Update MP-CNN to use torchtext for existing datasets: SICK and MSRVID.

I saved the torchtext Dataset subclasses at the root `/datasets` directory, so others can also use them for their models if needed.

Note, one may encounter some import errors for the SICK or MSRVID dataset modules. To resolve this one needs to make sure the repo root is in the `PYTHONPATH` environment variable. For example, for me this is `/u/z3tu/castorini/Castor`.

Addressed in latest update: ~~**Known bug**: this code cannot currently run on CPU, it must use GPU. Although PyTorch [documentation](http://pytorch.org/docs/master/cuda.html#torch.cuda.device) states that when `torch.cuda.device(idx)` is given negative `idx`, this is a no-op, when I give -1 for CPU mode, the code inside the context still gets run with the GPU selected. There are easy ways around this but I would like to check the best solution - and CPU is slow to train so I think this is fine for now.~~

Some performance measures:

**SICK**
Command: `python main.py mpcnn.sick.model.castor --dataset sick --epochs 19 --epsilon 1e-7 --dropout 0`

*Before torchtext*: Pearson's r - 0.8763
*After torchtext*: Pearson's r - 0.868353

**MSRVID**
Command: `python main.py mpcnn.msrvid.model.castor --dataset msrvid --batch-size 16 --epsilon 1e-7 --epochs 32 --dropout 0 --regularization 0.0025`

*Before torchtext*: Pearson's r - 0.9050
*After torchtext*: Pearson's r - 0.891191

I haven't tuned hyperparameters after the torchtext refactoring, so we can expect to get further improvements with more tuning.

@HTAustin @Impavidity please review?

### CODE REVIEW

File: datasets/idf_utils.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    """
+    Get pairwise word to document frequency.
+    For index i, if sentence i in sent_list_1 and sentence i in sent_list_2 both
+    container word w, then w is counted only once.
+    Returns a dictionary mapping words to number of sentence pairs the word appears in.
+    """
+    word_to_doc_cnt = defaultdict(int)
+
+    sent_list_2_iter = iter(sent_list_2)
+    for s1 in sent_list_1:
```
Comment (HTAustin) [inline datasets/idf_utils.py]: Why not using `for s1,s2 in zip(sent_list_1,sent_list_2): ` to loop two lists.

File: mp_cnn/main.py [hunk 2]
```diff
@@ -50,35 +51,38 @@
...
+        = MPCNNDatasetFactory.get_dataset(args.dataset, args.word_vectors_dir, args.word_vectors_file, args.batch_size, args.device)
 
     filter_widths = list(range(1, args.max_window_size + 1)) + [np.inf]
-    input_channels = 300
-    model = MPCNN(input_channels, args.holistic_filters, args.per_dim_filters, filter_widths, args.hidden_units, train_loader.dataset.num_classes, args.dropout, args.sparse_features)
-    if args.cuda:
+    model = MPCNN(embedding, args.holistic_filters, args.per_dim_filters, filter_widths,
+                    args.hidden_units, dataset_cls.NUM_CLASSES, args.dropout, args.sparse_features)
+
+    with torch.cuda.device(args.device):
```
Comment (Impavidity) [inline mp_cnn/main.py]: Why not just condition of args.device ? If it equals -1, then we do not move model to gpu.

File: datasets/msrvid.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        class_probs[floor] = ceil - sim
+        class_probs[ceil] = sim - floor
+
+    return class_probs
+
+
+class MSRVID(Dataset):
+    NAME = 'msrvid'
+    NUM_CLASSES = 6
+    ID_FIELD = Field(sequential=False, tensor_type=torch.FloatTensor, use_vocab=False, batch_first=True)
```
Comment (HTAustin) [inline datasets/msrvid.py]: ID's type should be `IntTensor` right?

File: mp_cnn/model.py [hunk 4]
```diff
@@ -22,20 +22,20 @@ def __init__(self, n_word_dim, n_holistic_filters, n_per_dim_filters, filter_wid
                 continue
 
             holistic_conv_layers.append(nn.Sequential(
-                nn.Conv1d(n_word_dim, n_holistic_filters, ws),
+                nn.Conv1d(self.n_word_dim, n_holistic_filters, ws),
```
Comment (HTAustin) [inline mp_cnn/model.py]: Also add `self` for other parameters such as `n_holistic_filters`.

File: mp_cnn/model.py [hunk 5]
```diff
@@ -61,24 +61,25 @@ def _get_blocks_for_sentence(self, sent):
...
+                    'mean': F.avg_pool1d(sent_flattened, sent_flattened_size).view(sent.size(0), -1)
                 }
                 continue
 
             holistic_conv_out = self.holistic_conv_layers[ws - 1](sent)
             block_a[ws] = {
-                'max': F.max_pool1d(holistic_conv_out, holistic_conv_out.size(2)).view(-1, self.n_holistic_filters),
-                'min': F.max_pool1d(-1 * holistic_conv_out, holistic_conv_out.size(2)).view(-1, self.n_holistic_filters),
-                'mean': F.avg_pool1d(holistic_conv_out, holistic_conv_out.size(2)).view(-1, self.n_holistic_filters)
+                'max': F.max_pool1d(holistic_conv_out, holistic_conv_out.size(2)).contiguous().view(-1, self.n_holistic_filters),
```
Comment (HTAustin) [inline mp_cnn/model.py]: What is the advantage of `contiguous()`

File: mp_cnn/main.py [hunk 6]
```diff
@@ -97,7 +101,7 @@
         trainer.train(args.epochs)
 
     model = torch.load(args.model_outfile)
-    test_evaluator = MPCNNEvaluatorFactory.get_evaluator(args.dataset, model, test_loader, args.batch_size, args.cuda)
+    test_evaluator = MPCNNEvaluatorFactory.get_evaluator(dataset_cls, model, test_loader, args.batch_size, args.device)
```
Comment (HTAustin) [inline mp_cnn/main.py]: Duplication with line 80.

File: mp_cnn/main.py (hunk 2 above)
Comment (tuzhucheng) [reply mp_cnn/main.py]: Yeah your suggestion is the way to go. I was just concerned about too many `if` checks in `train.py` since code doesn't look clean

File: datasets/msrvid.py (hunk 3 above)
Comment (tuzhucheng) [reply datasets/msrvid.py]: Good catch, should be `LongTensor` (there's no int in PyTorch)

File: datasets/idf_utils.py (hunk 1 above)
Comment (tuzhucheng) [reply datasets/idf_utils.py]: Nice, forgot about it

File: datasets/msrvid.py (hunk 3 above)
Comment (tuzhucheng) [reply datasets/msrvid.py]: You're right, there is `IntTensor` - I didn't recall seeing it before

File: mp_cnn/model.py (hunk 5 above)
Comment (Impavidity) [reply mp_cnn/model.py]: This is needed in this case (before view)

File: mp_cnn/main.py (hunk 6 above)
Comment (tuzhucheng) [reply mp_cnn/main.py]: It looks the same but is meant to be duplicate. This line creates a `test_evaluator` from a saved model on disk whereas the previous one is used for training.

File: mp_cnn/model.py (hunk 4 above)
Comment (tuzhucheng) [reply mp_cnn/model.py]: `n_holistic_filters` is a parameter of the function whereas `n_word_dim` is no longer. `n_holistic_filters` is the same as `self.n_holistic_filters` but more concise.

**Comment by tuzhucheng** [comment]
@Victor0118 please feel free to review too if you have time

File: mp_cnn/dataset.py [hunk 7]
```diff
@@ -26,193 +24,38 @@
 logger.addHandler(ch)
 
 
-class DatasetType(Enum):
-    TRAIN = 1
-    TEST = 2
-    DEV = 3
+unknown_vec = None
```
Comment (Victor0118) [inline mp_cnn/dataset.py]: Why do you need unknown_vec here? I am a little confused.

Comment (tuzhucheng) [reply mp_cnn/dataset.py]: I want to use the same unknown vector for all OOV words' vectors. In torchtext I must pass in a function that when given a tensor `t`, return a tensor of the same size as `t`. In the function below I fix a vector the first the function is run so for later runs I get the same vector.

```python
 +def unk(tensor):
 +    global unknown_vec
 +    if unknown_vec is None:
 +        unknown_vec = torch.Tensor(tensor.size())
 +        unknown_vec.normal_(0, 0.01)
 +    return unknown_vec
```

Please let me know if there is a better way.

Comment (Victor0118) [reply mp_cnn/dataset.py]: I see. It is reasonable if you want to use a static variable in python like this. But people prefer to avoid using global variable in python. https://stackoverflow.com/questions/279561/what-is-the-python-equivalent-of-static-variables-inside-a-function may provide a better solution if you do not share a global variable with other modules.

Comment (tuzhucheng) [reply mp_cnn/dataset.py]: I will fix this in the next update to the PR

**Comment by tuzhucheng** [comment]
Updated version can run on CPU too

**Comment by Victor0118** [comment]
LGTM good job

**Comment by HTAustin** [comment]
LGTM.

File: datasets/msrvid.py [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+        class_probs[ceil] = sim - floor
+
+    return class_probs
+
+
+class MSRVID(Dataset):
+    NAME = 'msrvid'
+    NUM_CLASSES = 6
+    ID_FIELD = Field(sequential=False, use_vocab=False, batch_first=True)
+    TEXT_FIELD = Field(batch_first=True, tokenize=lambda x: x)  # tokenizer is identity since we already tokenized it to compute external features
```
Comment (rosequ) [inline datasets/msrvid.py]: Would a better way be `sequential = False`?

Comment (tuzhucheng) [reply datasets/msrvid.py]: Hmm, a text field (a series of words) is sequential though - so not sure if that will be confusing so I used an identity tokenize function.
